cpp/include/cudf/ast/detail/operators.hpp

```diff
@@ -780,6 +789,51 @@ struct operator_functor<ast_operator::NOT, false> {
   }
 };
 
+// Type traits used to determine when a casting operation is valid.
+template <typename T, typename U, typename = void>
+struct can_cast : std::false_type {
+};
+
+template <typename T, typename U>
+struct can_cast<T, U, std::void_t<decltype(static_cast<U>(std::declval<T>()))>> : std::true_type {
+};
+
+template <typename T, typename U>
+constexpr bool can_cast_v = can_cast<T, U>::value;
```

Comment:
how about using `std::is_convertible_v<From, To>`?
https://en.cppreference.com/w/cpp/types/is_convertible